cpp/src/arrow/dataset/file_base.cc

```diff
@@ -81,6 +81,18 @@ Result<std::shared_ptr<io::RandomAccessFile>> FileSource::Open() const {
   return custom_open_();
 }
 
+Future<std::shared_ptr<io::RandomAccessFile>> FileSource::OpenAsync() const {
+  if (filesystem_) {
+    return filesystem_->OpenInputFileAsync(file_info_);
+  }
+  
+  if (buffer_) {
+    return Future<std::shared_ptr<io::RandomAccessFile>>::MakeFinished(std::make_shared<io::BufferReader>(buffer_));
+  }
+
+  return Future<std::shared_ptr<io::RandomAccessFile>>::MakeFinished(custom_open_());
```

Comment:
Oh, since `custom_open_` might still blocking here?

Reply:
Yes, wasn't sure how to deal with it. FileSource would need its own `IOContext` instance to run it async?

Reply:
You can leave a comment here, and keep it here? I think at lease, open via fileSystem could be help, and `custom_open_` won't be worse

Reply:
👍 left a todo

Reply:
I think this TODO is valuable but should be linked to a new issue where we extend `custom_open_` to return `Future<std::share_ptr<io::RandomAccessFile>>`. That would be necessary to enable a custom but non blocking open

Reply:
Created a new issue and updated the comment https://github.com/apache/arrow/issues/37962